### PR TITLE
Add support for placeholders and their substitutions in paths

### PR DESCRIPTION
**Pull request by adellhk**
Using a string at DLD instance instantiation, users can configure one placeholder's replacement with that string
```
public BaseDeepLinkDelegate(
    List<? extends BaseRegistry> registries,
    String pathVariableReplacementValue
  ) {
    this.registries = registries;
    this.pathVariableReplacementValue = pathVariableReplacementValue;
  }
```

Given: 
- `@DeepLink("https://www.example.com/-replaceable-path-variable-/")`
-`DeepLinkDelegate(registries, "obamaOs")`

Then:
- `DeepLinkDelegate.supportsUri("https://www.example.com/obamaOs/") == true`

@rossbacher @BenSchwab

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,4 +15,7 @@ lib
 .settings
 .DS_Store
 local.properties
-.gradle
\ No newline at end of file
+.gradle
+
+# deepLinkDispatch
+*.idx
```
Comment (rossbacher) [inline .gitignore]: We should not ignore all .idx files. should try `dld_match_*.idx`

File: deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkUri.java [hunk 2]
```diff
@@ -386,7 +390,48 @@ Builder newBuilder() {
   public static DeepLinkUri parse(String url) {
```
Comment (rossbacher) [inline deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkUri.java]: Maybe add `@Nullable` annotation for the return value.

File: deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkUri.java [hunk 3]
```diff
@@ -386,7 +390,48 @@ Builder newBuilder() {
...
+   * @return URI or null if no replacements could be made.
+   */
+  public static DeepLinkUri parseWithReplacement(String pathVariableReplacementValue) {
+    if (lastBuilder == null
+      || pathVariableReplacementValue == null
+      || pathVariableReplacementValue.isEmpty()
+    ) {
+      return null;
+    }
+    List<String> segments = lastBuilder.encodedPathSegments;
```
Comment (rossbacher) [inline deeplinkdispatch-base/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkUri.java]: So we are expecting that this is in a certain state (and not the instance but the class)?

File: sample/src/main/java/com/airbnb/deeplinkdispatch/sample/MainActivity.java [hunk 4]
```diff
@@ -28,7 +28,13 @@
 import androidx.core.app.ActivityCompat;
 import androidx.core.app.TaskStackBuilder;
 
-@DeepLink({ "dld://classDeepLink", "http://example.com/foo{arg_end}", "http://example.com/{arg_start}bar", "dld://example.com/deepLink" })
+@DeepLink({
+  "dld://classDeepLink",
+  "http://example.com/foo{arg_end}",
+  "http://example.com/{arg_start}bar",
+  "dld://example.com/deepLink",
+  "https://www.example.com/-replaceable-path-variable-/bar"
```
Comment (rossbacher) [inline sample/src/main/java/com/airbnb/deeplinkdispatch/sample/MainActivity.java]: Didn't we say we use something like `###somevalue###` for the placeholder? `-replaceable-path-variable-` is still to easy to be matched by something else.

File: sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java [hunk 5]
```diff
@@ -167,4 +170,11 @@ public void testSupportsUri() throws Exception {
     assertThat(deepLinkDelegate.supportsUri("dld://classDeepLink"), equalTo(true));
     assertThat(deepLinkDelegate.supportsUri("some://weirdNonExistentUri"), equalTo(false));
   }
+
+  @Test
```
Comment (rossbacher) [inline sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java]: 👍

File: deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.java [hunk 6]
```diff
@@ -23,25 +23,64 @@
...
+   */
+  protected final String pathVariableReplacementValue;
 
   public List<? extends BaseRegistry> getRegistries() {
     return registries;
   }
 
   public BaseDeepLinkDelegate(List<? extends BaseRegistry> registries) {
     this.registries = registries;
+    this.pathVariableReplacementValue = null;
```
Comment (rossbacher) [inline deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.java]: You can just add that to definition.

File: deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.java [hunk 7]
```diff
@@ -23,25 +23,64 @@
...
-        break;
+        return entryIdxMatch;
       }
     }
-    return entryIdxMatch;
+
+    // If we didn't find a match, try again with substituting pathVariableReplacementValue.
+    // Since optimized idxMatch is so absurdly fast, this will have no impact on overall
+    // performance.
+    if (entryIdxMatch == null) {
```
Comment (rossbacher) [inline deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.java]: This is assuming static state in the `DeeplinkUri` class, need to remove that.

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 8]
```diff
@@ -422,16 +423,29 @@ private static PackageElement getPackage(Element type) {
...
     for (int i = 0; i < totalElements; i++) {
-      registriesInitializer.add("$L$L",
-          decapitalize(moduleNameToRegistryName(registryClasses.get(i))),
-          i < totalElements - 1 ? ",\n" : "\n");
+      moduleRegistriesArgument.add("$L$L",
+        decapitalize(moduleNameToRegistryName(registryClasses.get(i))),
+        i < totalElements - 1 ? ",\n" : "");
     }
 
+    CodeBlock registriesInitializerBuilder = CodeBlock.builder()
```
Comment (rossbacher) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: Could you add some comment of what this is doing, it's so hard to read.

**Comment by adellhk** [comment]
closing in favor of https://github.com/airbnb/DeepLinkDispatch/pull/274

**Pull request closed without merging**